service/storage_proxy.cc

```diff
@@ -3659,9 +3659,7 @@ storage_proxy::mutate_atomically_result(std::vector<mutation> mutations, db::con
 
         future<result<>> send_batchlog_mutation(mutation m, db::consistency_level cl = db::consistency_level::ONE) {
             return _p.mutate_prepare<>(std::array<mutation, 1>{std::move(m)}, cl, db::write_type::BATCH_LOG, _permit, [this] (const mutation& m, db::consistency_level cl, db::write_type type, service_permit permit) {
-                auto& table = _p._db.local().find_column_family(_schema->id());
-                auto ermp = table.get_effective_replication_map();
-                return _p.create_write_response_handler(std::move(ermp), cl, type, std::make_unique<shared_mutation>(m), _batchlog_endpoints, {}, {}, _trace_state, _stats, std::move(permit), std::monostate(), is_cancellable::no);
+                return _p.create_write_response_handler(_ermp, cl, type, std::make_unique<shared_mutation>(m), _batchlog_endpoints, {}, {}, _trace_state, _stats, std::move(permit), std::monostate(), is_cancellable::no);
```

Comment:
Looking at all those patches I wonder if we should change `inet_address_vector_replica_set` to hold ermp.

Reply:
It could be self contained also with a token_metadata_ptr (that holds a stable topology object) and a vector of `node*` of nodes in that topology instance.
From that you can get their address, dc/rack location, host_id, etc.

Reply:
an effective_replication_map_ptr would also provide the same stable topology, but I'm not sure it always makes sense for all uses of `inet_address_vector_replica_set`

Reply:
> The lifetime of ermp starts earlier than the replica set and ends later.

It does not mean it cannot contain a pointer to erm that was used to create it.

Reply:
It can, but it doesn't help anything to ensure we use exactly one ermp through the lifetime of a query.

Reply:
> It can, but it doesn't help anything to ensure we use exactly one ermp through the lifetime of a query.

At least for the last 3 or 4 Benny's PRs it would have helped. Of course nothing prevents someone to ignore an erm from the  `inet_address_vector_replica_set` and obtain another one. But nothing prevents it if you pass it separately as well.

Reply:
It's just duplication. You have to create it before you get the replica set, and (usually) you have to keep it after you get rid of it.

Reply:
You do not need to create it explicitly. `erm::get_natural_endpoints` and friends can just return it.

Reply:
But you need erm before get_live_endpoints and friends. For example, you start with obtaining the replication strategy.

Reply:
I do not understand. Here is the code example:

```
    auto erm = table.get_effective_replication_map();
    inet_address_vector_replica_set natural_endpoints = erm->get_natural_endpoints_without_node_being_replaced(token);
    inet_address_vector_topology_change pending_endpoints = erm->get_pending_endpoints(token);

```

Here the returned value can contain the reference to the erm it was obtained through. The erm can be used for other things as well. It has its own lifetime.

Reply:
It's just one example. In other cases you need erm before.

Why have multiple copies of the same thing?

Reply:
erm is a shared pointer. It is designed to have multiple copies. When we pass it down the call chain we also pass a copy.

Reply:
Yes, but if we can avoid duplicates, it's better.

Copies make sense when the lifetime diverges (we don't have one path that strictly contains the other). But here, the write handler outlives everything else.

Reply:
> Yes, but if we can avoid duplicates, it's better.

But we do not. We pass a copy down the call.

> 
> Copies make sense when the lifetime diverges (we don't have one path that strictly contains the other). But here, the write handler outlives everything else.

The code we are discussing is the code that creates write handlers, so you cannot say that a write handler outlives it. Of course there is no point to hold erm twice in the write handler (once inside erm and second time in a replica set structire) if that what you mean. The outer one can be dropped. I am not saying there is a lot of value in it now, when all the code is written already. but if it would have been this way from the start we would not have seen those kind of errors.